### ISSUE TITLE
Frequent disconnections from RDP with TS Gateway

### ISSUE THREAD

**Issue by nalimilan** [FreeRDP/FreeRDP#2431]
So recent changes have made me able to connect using TS Gateway (https://github.com/FreeRDP/FreeRDP/issues/1589), which is great. Now I'm going to complain about something else (you know how people are... :-): I get quite frequent disconnections, and even when everything works error messages are printed.

```
$ xfreerdp /v:CLUSTER /d:DOMAIN /u:USER /p:PASSWD /g:GW /gd:DOMAIN /gu:USER /gp:PASSWD
[17:18:48:074] [24889:24890] [INFO][com.freerdp.core.gateway.tsg] - TS Gateway Connection Success
[17:19:03:292] [24889:24890] [ERROR][com.freerdp.core.gateway.rts] - Receiving FlowControlAckWithDestination RTS PDU: BytesReceived: 33230 AvailableWindow: 65056
[17:19:11:193] [24889:24890] [INFO][com.freerdp.client.x11] - Network disconnect!
[17:19:11:193] [24889:24890] [ERROR][com.freerdp.client.x11] - Failed to check FreeRDP file descriptor
[17:19:11:193] [24889:24890] [ERROR][com.freerdp.core.gateway.tsg] - rpc_write failed!
```

iTap suffers from some instability too, so maybe these are only network problems. Feel free to close if that's the case. If there's something more, please ask if you need more info.

**Comment by bmiklautz** [comment]
@nalimilan you tested with recent master? What kind of connection do you have to your server? Low bandwith or hight delay?

**Comment by nalimilan** [comment]
Yes, this is with master as of yesterday. I tested it with various connections, I think they are reasonably fast (between two research institutes in Paris, or from my home cable connection to another institute). Is there any way to check the quality of the connection?

**Comment by nalimilan** [comment]
FWIW, it also happens from a LAN where the Windows RDP client never suffers from connection problems.

**Comment by dbungert** [comment]
@nalimilan What OS Gateway Server / Desktop are you connecting to?

**Comment by nalimilan** [comment]
I'm not sure what kind of useful information I can give you. That's a Windows Server 2008 R2 Enterprise machine.

**Comment by dbungert** [comment]
I'm able to reproduce this issue with master, and I see very similar (if not identical) result based on our fork from freerdp-1.1.  The test case failing for me with master is using 2012r2 fully patched for the Gateway, and 8.1 fully patched for the Desktop.  The test environment is also a very favorable LAN configuration - test device is on the same switch as the HyperV server, and the VMs for AD, GW, Desktop are all in HyperV.  So network dropouts are unlikely to be the culprit.

**Comment by dbungert** [comment]
Retesting with the new http transport implementation, I seem to see the same issue there.

[13:04:17:572] [27770:27771] [ERROR][com.freerdp.locale] - Unable to find a ma
tch for unix timezone: Etc/UTC
[13:12:16:627] [27770:27771] [WARN][com.freerdp.core.gateway.rdg] - Unknown Co
ntrol Packet received: D
[13:12:21:908] [27770:27771] [INFO][com.freerdp.client.x11] - Network disconne
ct!
[13:12:21:909] [27770:27771] [ERROR][com.freerdp.client.x11] - Failed to check
 FreeRDP file descriptor

**Comment by dbungert** [comment]
A pair of fixes for Gateway over RPC were merged today. Give master a try.

**Comment by nalimilan** [comment]
From a quick test, I still get the same disconnection. :-/

**Comment by chrisdavis925** [comment]
I can confirm that the disconnection is still present after x amount minutes. . :(

Any idea how to resolve/fix?

**Comment by chrisdavis925** [comment]
Just to provide more info: I built master this morning and connected to a 2012 R2 VDI host which has a pool of Windows 7 VMs. Windows 7 VMs are using RDP 8. 

Command used to connect: 

Log from a session that dissconnected after a few minutes:
 xfreerdp /v:SERVER-HYPERV /u:Administrator /p:password /d:domain.local /g:remote.domain.com /cert-ignore /load-balance-info:tsv://VMResource.1.test /gt:rpc

[10:58:16:221] [6124:6125] [INFO][com.freerdp.crypto] - creating directory /root/.config/freerdp
[10:58:16:222] [6124:6125] [INFO][com.freerdp.crypto] - creating directory /root/.config/freerdp/certs
[10:58:16:222] [6124:6125] [INFO][com.freerdp.crypto] - creating directory /root/.config/freerdp/server
[10:58:16:222] [6124:6125] [ERROR][com.freerdp.crypto] - certificate_store_open: error opening [/root/.config/freerdp/known_hosts] for writing
[10:58:16:294] [6124:6125] [INFO][com.freerdp.crypto] - creating directory /root/.config/freerdp
[10:58:16:294] [6124:6125] [INFO][com.freerdp.crypto] - creating directory /root/.config/freerdp/certs
[10:58:16:294] [6124:6125] [INFO][com.freerdp.crypto] - creating directory /root/.config/freerdp/server
[10:58:16:294] [6124:6125] [ERROR][com.freerdp.crypto] - certificate_store_open: error opening [/root/.config/freerdp/known_hosts] for writing
[10:58:17:937] [6124:6125] [INFO][com.freerdp.core.gateway.tsg] - TS Gateway Connection Success
[10:58:17:038] [6124:6125] [INFO][com.freerdp.crypto] - creating directory /root/.config/freerdp
[10:58:17:038] [6124:6125] [INFO][com.freerdp.crypto] - creating directory /root/.config/freerdp/certs
[10:58:17:038] [6124:6125] [INFO][com.freerdp.crypto] - creating directory /root/.config/freerdp/server
[10:58:17:038] [6124:6125] [ERROR][com.freerdp.crypto] - certificate_store_open: error opening [/root/.config/freerdp/known_hosts] for writing
[10:58:19:650] [6124:6125] [INFO][com.freerdp.crypto] - creating directory /root/.config/freerdp
[10:58:19:651] [6124:6125] [INFO][com.freerdp.crypto] - creating directory /root/.config/freerdp/certs
[10:58:19:651] [6124:6125] [INFO][com.freerdp.crypto] - creating directory /root/.config/freerdp/server
[10:58:19:651] [6124:6125] [ERROR][com.freerdp.crypto] - certificate_store_open: error opening [/root/.config/freerdp/known_hosts] for writing
[10:58:19:728] [6124:6125] [INFO][com.freerdp.crypto] - creating directory /root/.config/freerdp
[10:58:19:728] [6124:6125] [INFO][com.freerdp.crypto] - creating directory /root/.config/freerdp/certs
[10:58:19:728] [6124:6125] [INFO][com.freerdp.crypto] - creating directory /root/.config/freerdp/server
[10:58:19:728] [6124:6125] [ERROR][com.freerdp.crypto] - certificate_store_open: error opening [/root/.config/freerdp/known_hosts] for writing
[10:58:19:368] [6124:6125] [INFO][com.freerdp.core.gateway.tsg] - TS Gateway Connection Success
[10:58:19:471] [6124:6125] [INFO][com.freerdp.crypto] - creating directory /root/.config/freerdp
[10:58:19:471] [6124:6125] [INFO][com.freerdp.crypto] - creating directory /root/.config/freerdp/certs
[10:58:19:471] [6124:6125] [INFO][com.freerdp.crypto] - creating directory /root/.config/freerdp/server
[10:58:19:471] [6124:6125] [ERROR][com.freerdp.crypto] - certificate_store_open: error opening [/root/.config/freerdp/known_hosts] for writing
[11:04:34:493] [6124:6125] [INFO][com.freerdp.client.x11] - Network disconnect!
[11:04:34:493] [6124:6125] [ERROR][com.freerdp.client.x11] - Failed to check FreeRDP file descriptor

**Comment by chrisdavis925** [comment]
Just to provide another update/insight:

-Connection will fail to my 2012R2 server environment through a gateway brokered to VMs on a hyperV host if you supply the /gu /gp /gd flags. Even though they are the same as the login credentials. I would think it should still connect if you provide these details, but it fails. It works if you don't provide the /gu details, but the session will disconnect randomly after a few moments. This only works when /gt:rpc is specified. 

-If you try to connect with /sec:rdp instead of /sec:nla I get the error "Failed to check FreeRDP file descriptor". This is the same error that is seen after a random disconnect occurs. Perhaps there is something going on where the /nla setting is getting reverted after the connection starts? Just guessing at this point since the errors are the same.

**Comment by dbungert** [comment]
@chrisdavis925 no new suggestions, sorry.  I'm setting up a local server intended for long-term gateway connection testing, that may help me find more resolutions for this class of defect.

**Comment by chrisdavis925** [comment]
Thank you @dbungert . I am happy to provide you access to a remote server as well for testing, if it helps. 

I also want to confirm that load-balance-info broker redirection works just fine on LAN without gateway redirection - so the problem appears to be specific to the gateway feature.

**Comment by charlesg** [comment]
Experience the same issue. When I first tested over LAN, I noticed it once or twice but overall it was pretty stable. Now over WAN, it's unusable. The connection comes up half the times and most times it closes with a timeout before I get to login.

**Comment by chrisdavis925** [comment]
@dbungert  have you had any luck with researching this more?
I really appreciate all of your efforts. :)

**Comment by varun580** [comment]
Faced the same error "Failed to check file descriptor"  xfreerdp closing the connection within 2secs.
 Adding option  +auto-reconnect to the command  worked for me, this option makes an immediate suto reconnect as soon as the connection closes.

Eg command:  xfreerdp /u:varun /p:pass /g:<gateway-ip> /gd:<domaiin> /gu:varun /gp:password /sec:rdp +auto-reconnect  /f

[10:28:40:291] [18851:-1477536000] [INFO][com.freerdp.core.gateway.tsg] - TS Gateway Connection Success
[10:34:38:842] [18851:-1477536000] [INFO][com.freerdp.client.x11] - Network disconnect!
[10:34:38:842] [18851:-1477536000] [INFO][com.freerdp.client.x11] - Attempting reconnect (1 of 20)

**Comment by chrisdavis925** [comment]
Just checking on this again - anyone have any update? Would love to see a fix for this, but I don't have the knowledge on how to fix it. 

I super appreciate the dev support on this issue :)

**Comment by ithh** [comment]
So this problem went away for a while and recently came back. I was connecting through a Windows 2008R2 Gateway server without issues.

It was recently upgraded to 2012R2 and I was not able to connect at all at first. I finally noticed I had to force the gateway transport: /gt:rpc otherwise it won't connect (ERR_CONNECT_TRANSPORT_FAILED).

With RPC I connect okay but I get dropped after a minute or two:

- Failed to check FreeRDP file descriptior
- rpc_write failed!

I just pulled the current master. Is that a new issue with 2012R2 or is this part of this same disconnect issue?

**Comment by chrisdavis925** [comment]
I have been testing and building latest builds from Master and have always experienced the "Failed to check FreeRDP file descriptor" when connecting to a Windows 7 Desktop via Windows Server 2012 R2 Gateway. This typically happens within a few minutes of connecting.

**Comment by chrisdavis925** [comment]
It appears that the most recent master fixed the problems with 2012 R2 gateway. 
It seems to me like there might be a slow memory leak with this version, though.

**Comment by chrisdavis925** [comment]
I spoke too soon. It seems to disconnect a lot less, but it still disconnects with the same file descriptor error.

**Comment by hardening** [comment]
@chrisdavis925 we're aware of an issue with TSG + Tls security, it doesn't happen when using RDP security. Would you mind give a try ?

**Comment by chrisdavis925** [comment]
@hardening - I am using /gt:rpc , which is the only way the connection works. The disconnects still happen with master pulled yesterday. I will say it appears to disconnect less, but it does still disconnect with the file descriptor error.

**Comment by Faflu** [comment]
My experience is that as long as I do something in the remote desktop, there are no isssues. Only after some period of inactivity I experience the same issue. The gateway and TS are on Win2k12. FreeRDP is compiled from sources pulled from git.

**Comment by ipeacocks** [comment]
Have you tried another colour depth? For example 16?

**Comment by akallabeth** [comment]
still an issue?

**Comment by nalimilan** [comment]
I haven't seen it in a while, so looks like it's fixed now. Thanks!
